### PR TITLE
Provide an explicit default CMAKE_BUILD_TYPE

### PR DESCRIPTION
**Pull request by masterleinad**
Without this, we don't set any debug or optimization flags by default.
The actual default type is debatable but I think `Release` is an appropriate choice.

### CODE REVIEW

**Comment by crtrott** [comment]
I assume this will not apply to Trilinos since its inside the not having set Project Name clause?

**Comment by masterleinad** [comment]
> I assume this will not apply to Trilinos since its inside the not having set Project Name clause?

Yes, that was the intention. I think it's not our business to do anything if `Kokkos` is just a subproject.

**Review by dalg24 — APPROVED** [review]

File: CMakeLists.txt [hunk 1]
```diff
@@ -83,6 +83,13 @@ IF(NOT KOKKOS_HAS_TRILINOS)
   ENDif()
   IF(NOT DEFINED ${PROJECT_NAME})
```
Comment (jjwilke) [inline CMakeLists.txt:84]: Shouldn't this be 
````
IF (NOT DEFINED PROJECT_NAME)
````
?

File: CMakeLists.txt [hunk 2]
```diff
@@ -83,6 +83,13 @@ IF(NOT KOKKOS_HAS_TRILINOS)
...
     PROJECT(Kokkos CXX)
+    IF (NOT CMAKE_BUILD_TYPE)
```
Comment (jjwilke) [inline CMakeLists.txt:86]: Would this better to check `DEFINED CMAKE_BUILD_TYPE`?  There are uses for this being empty (i.e. CMake adds none of its own flags).

File: CMakeLists.txt [hunk 3]
```diff
@@ -83,6 +83,13 @@ IF(NOT KOKKOS_HAS_TRILINOS)
...
+      SET(DEFAULT_BUILD_TYPE "Release")
+      MESSAGE(STATUS "Setting build type to '${DEFAULT_BUILD_TYPE}' as none was specified.")
+      SET(CMAKE_BUILD_TYPE "${DEFAULT_BUILD_TYPE}" CACHE STRING
+          "Choose the type of build, options are: Debug, Release, RelWithDebInfo and MinSizeRel."
```
Comment (jjwilke) [inline CMakeLists.txt:90]: Same question. If we only go here if build type has not been defined, we can just use a regular variable and don't need a cache variable.

**Review by jjwilke — CHANGES_REQUESTED** [review]

File: CMakeLists.txt (hunk 1 above)
Comment (jjwilke) [reply CMakeLists.txt:84]: I realize this line already exists, but I think this is wrong.  This line is 3 years old and somehow never go caught.

File: CMakeLists.txt (hunk 2 above)
Comment (masterleinad) [reply CMakeLists.txt:86]: The purpose of this pull request was to default to something reasonable when no build type is provided when calling `cmake`.
If I check
```
IF(NOT DEFINED CMAKE_BUILD_TYPE)
```
I still get an empty build type in that case. In other words, `CMAKE_BUILD_TYPE` is already defined (as empty) at this point.

File: CMakeLists.txt (hunk 3 above)
Comment (masterleinad) [reply CMakeLists.txt:90]: This needs to be a `CACHE` variable to have any effect.

Comment (jjwilke) [reply CMakeLists.txt:90]: Touché.  This does get automatically defined in the CACHE even if not given on the command line. My bad.

File: CMakeLists.txt (hunk 2 above)
Comment (jjwilke) [reply CMakeLists.txt:86]: Yep. I see the problem. See comment below.

File: CMakeLists.txt (hunk 1 above)
Comment (jjwilke) [reply CMakeLists.txt:84]: This is obviously unrelated to the PR, per se, but should we fix it now?

**Review by jjwilke — APPROVED** [review]

**Review by dalg24 — DISMISSED** [review]
Everything else look good.

File: CMakeLists.txt [hunk 4]
```diff
@@ -81,8 +81,15 @@ IF(NOT KOKKOS_HAS_TRILINOS)
       SET(ENV{CXX} ${SPACK_CXX})
     ENDIF()
   ENDif()
-  IF(NOT DEFINED ${PROJECT_NAME})
+  IF(NOT DEFINED PROJECT_NAME)
     PROJECT(Kokkos CXX)
+    IF (NOT CMAKE_BUILD_TYPE)
+      SET(DEFAULT_BUILD_TYPE "Release")
```
Comment (dalg24) [inline CMakeLists.txt:87]: Please change to `RelWithDebInfo`

**Pull request merged**